src/phoenix/server/api/queries.py

```diff
@@ -1702,6 +1721,112 @@ async def get_project_session_by_id(
             return ProjectSession(id=session_row.id, db_record=session_row)
         return None
 
+    @strawberry.field
+    async def apply_chat_template(
+        self,
+        info: Info[Context, None],
+        template: list[ChatCompletionMessageInput],
+        template_options: PromptTemplateOptions,
+    ) -> PromptChatTemplate:
+        """
+        Applies template formatting to chat completion messages.
+
+        Takes a list of messages with template placeholders and template options
+        (format and variables), and returns the messages with placeholders replaced.
+        """
+        formatter = get_template_formatter(template_options.format)
+        # Ensure variables is a dict - JSON scalar can be any JSON type
+        raw_variables = template_options.variables
+        if isinstance(raw_variables, dict):
+            variables = raw_variables
+        elif isinstance(raw_variables, str):
+            parsed = json.loads(raw_variables)
+            if not isinstance(parsed, dict):
+                raise ValueError("Variables JSON string must parse to a dictionary")
+            variables = parsed
+        else:
+            raise ValueError("Variables must be a dictionary or a JSON string")
```

Comment:
it's odd to me that the JSON scalar is so odd, I can refactor this I suppose but I have tests for this case so might want to wait